### PR TITLE
[#30] Optimize the whole import deletion

### LINKED ISSUE #30: Optimize whole import deletion
Currently we just assume and length of line is 100 and we try to delete entry on every position. But this can be optimized to delete only from starting positions of entries:

https://github.com/kowainik/smuggler/blob/4fbaafe8c213790098b365bbd4163fa492d90a4e/src/Smuggler/Plugin.hs#L85

### PR DESCRIPTION
**Pull request by v0d1ch**
fixes #30

### CODE REVIEW

**Review by vrom911 — COMMENTED** [review]
Thank you very much for improvements!

Could you please check why tests are failing? You should probably be looking at this tricky test:

https://github.com/kowainik/smuggler/blob/master/test/Test/MultilineUnused.hs

(We actually need better test failure message 😃)

**Comment by vrom911** [comment]
_Side Note:_ I've just created https://github.com/kowainik/smuggler/issues/53 to improve tests output

**Comment by v0d1ch** [comment]
I have determined that the testing outputs don't match so that is why the test fails:
```
module Test.MultilineUnused where

import Data.Bool (bool)

foo = bool

VS

module Test.MultilineUnused where

import Data.Bool (bool) intercalate, replicate, reverse, take, takeWhile,

foo = bool
```
I will investigate further how to fix it, seems like my changes with columns do not work.

**Comment by v0d1ch** [comment]
One question. How do I debug the plugin a part from tests ? If I am right this source plugin is running after typechecking and I would like to trace some of my changes. I guess there is definitely a way to see some output, maybe I need to use some flag to dump it to terminal ? How do you debug while developing source plugins in general ?

**Comment by chshersh** [comment]
@v0d1ch Well, testing Source Plugins is a hard task. Especially for plugins that modify source code. To my knowledge, `smuggler` is the only plugin at the moment that modifies modules content automatically. The Source Plugins were created only recently, so there's no best-practices for that.

Since Source Plugins should be a part of compilation pipeline, you need to have some stanza which you will compile. So, usually I just put test case into `app/Main.hs` module and observe the result after compilation. But, since `smuggler` rewrites content of the file, it's not that convenient. For testing purposes there's implemented support for redirected output. If you add the following GHC option to the `executable` stanza — `-fplugin-opt=Smuggler.Plugin:out` — then the `smuggler` output will be redirected to file with the same path but with `.out` file extensions instead of `.hs` extension. So you can see the result of `smuggler` output without changing source code.

**Comment by chshersh** [comment]
@v0d1ch Also, we have this function for pretty-printing AST of Haskell Source code (or `ghc-exactprint` annotations):

* https://github.com/kowainik/smuggler/blob/7746b696f72ee6c0a844222552954c97b8be4408/src/Smuggler/Debug.hs#L21

It's commented in the separate module so we won't accidentally release it and dump redundant dependencies to users. But it's very useful for intermediate debug.

File: src/Smuggler/Plugin.hs [hunk 1]
```diff
@@ -89,10 +89,9 @@ unusedLocs (L (RealSrcSpan loc) decl, used, unused)
     = []
 
     -- Nothing used; drop entire decl
-    -- TODO: optimize
     | null used = [ (lineNum, colNum)
                   | lineNum <- [srcSpanStartLine loc .. srcSpanEndLine loc]
-                  , colNum <- [1..100]
+                  , colNum <-  [srcSpanStartCol loc .. srcSpanEndCol loc]
```
Comment (v0d1ch) [inline src/Smuggler/Plugin.hs]: I believe `srcSpanEndCol` here should give us the end column for the `SrcSpan`. Somehow it always yields `79` so I wonder if it somehow defaults to that. I asked about it on irc so let's see if someone has the answer. If I hardcode this value to say 100 everything works so I am not sure why this function does not give us the correct column end.

Comment (v0d1ch) [reply src/Smuggler/Plugin.hs]: Oh I found what was the problem. It picks up the last import line column. I'll need to read all line ends and then use the maximum.

File: test/Test.hs [hunk 2]
```diff
@@ -24,41 +22,12 @@ main = do
...
-            False <$ outputFailure testContent goldenContent
-
-    outputFailure :: String -> String -> IO ()
-    outputFailure given expected = do
-      boldMessage "Expected:"
-      putStrLn (importStatementsOnly expected)
-      boldMessage "But got:"
-      putStrLn (importStatementsOnly given)
-
-successCode, failureCode, resetCode :: [SGR]
```
Comment (chshersh) [inline test/Test.hs]: This is strange that all these functions are missing now :thinking: Probably some issues during conflicts resolution

**Comment by v0d1ch** [comment]
Yeah, no worries they will be back :)

File: src/Smuggler/Plugin.hs [hunk 3]
```diff
@@ -121,3 +119,6 @@ unusedLocs (L (RealSrcSpan loc) decl, used, unused)
         L _ (IEName (L (RealSrcSpan lieLoc) name)) <- [lieName]
         guard $ name `elem` unused
         pure (srcSpanStartLine lieLoc, srcSpanStartCol lieLoc)
+    getEndColMax u = maximum $ map (findColLoc . nameSrcSpan) u
+    findColLoc (RealSrcSpan l) =  srcSpanEndCol l
+    findColLoc (UnhelpfulSpan _) = 100
```
Comment (chshersh) [inline src/Smuggler/Plugin.hs]: Could you, pleaase, add types to functions inside `where`? This is very helpful for code understanding and maintaing, especially when working with complicated frameworks like GHC

File: src/Smuggler/Plugin.hs [hunk 4]
```diff
@@ -121,3 +119,6 @@ unusedLocs (L (RealSrcSpan loc) decl, used, unused)
         L _ (IEName (L (RealSrcSpan lieLoc) name)) <- [lieName]
         guard $ name `elem` unused
         pure (srcSpanStartLine lieLoc, srcSpanStartCol lieLoc)
+    getEndColMax u = maximum $ map (findColLoc . nameSrcSpan) u
```
Comment (chshersh) [inline src/Smuggler/Plugin.hs]: I'm a bit worried about the usage of partial function `maximum` but maybe the usage is completely correct in practice

**Review by chshersh — COMMENTED** [review]
@v0d1ch Thanks for the contribution! This is a good step towards improving the plugin! I left couple comments, but overall looks good :+1:

File: src/Smuggler/Plugin.hs (hunk 3 above)
Comment (v0d1ch) [reply src/Smuggler/Plugin.hs]: That is precisely what I argue about in work but my colleagues seem to think that type annotations are not so important inside of where. You are completely right about this.

File: src/Smuggler/Plugin.hs (hunk 4 above)
Comment (v0d1ch) [reply src/Smuggler/Plugin.hs]: I added custom function to calculate maximum list item or give a default value instead.

**Comment by v0d1ch** [comment]
@ChShersh && @vrom911 I am not too happy about this solution since it involves default value in case we run into `UnhelpfulSpan`. 

Then again, if the only way to calculate max column is from `[Name]` and  when converted to `SrcSpan` we do run into `UnhelpfulSpan` I don't know what else we could do. 

I guess if there is no better way of getting to single line column this solution would work, we just need to keep in mind that it will break if somebody adds import statement that is over 120 characters long. 
Maybe we want to bump this number to higher one but in practise I never add imports this long and I bet you don't either so this could be the corner case.

**Review by chshersh — APPROVED** [review]
@v0d1ch Thank you again for this helpful fix! Even if the solution is not the elegant one, it's still pretty good. Sometimes we need to choose between elegant solutions and pragmatic choices. I saw `UnhelpfulSpan` very few times in my life, so I don't expect to see any problems often. We can always adjust constant or even make it configurable but let's not overengineer until somebody really faces issues.

**Comment by v0d1ch** [comment]
I agree :)

**Pull request merged**